docs/source/clickhouse/schema_design.rst

```diff
@@ -72,3 +73,78 @@ Those styles of queries require a separate column to search.
+* (``widgets_aggregated_local``, ``widgets_aggregated_dist``) for the roll-up/aggregated (local, distributed) tables
+
+Materialized views are immutable so it's normal to have multiple versions of
+``widgets_aggregation_mv`` when behavior is updated, with suffixes like
+``widgets_aggregation_mv_v1``, ``widgets_aggregation_mv_v2``, etc. Migration
+between materialized view versions are described in the next section but in general
+old materialized views should be discarded once they are no longer used.
+
+Schema migrations using materialization_version
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As we discussed at the end of the prior section, materialized view logic cannot
+be updated in place. In order to continuously roll-up input data without data
+loss or duplication in the aggregate table, we control logic changes with a
+column on the raw table, ``materialization_version``, and making the materialized
+view logic depend on specific values of that column. To update MV logic, you
+create a new materialized view that looks for the last used value of
+``materialization_version`` plus one and then, after that's been created in all
+relevant environments, update the consumer to write the new materialization_version
+to the raw column.
+
+Here is how this might look in practice:
+
+Context:
+
+1. There is a raw table ``click_events_raw_local``, that has a field named
+   ``click_duration``, of type Float64. A snuba consumer is setting this to 0 for
+   certain types of click events.
+2. There is a materialized view ``click_events_aggregation_mv`` that is writing
+   a ``quantilesState()`` value for a ``click_duration`` column in ``click_events_aggregated_local``
+   including those zero-values. This materialized view looks for the value of
+   ``materialization_version = 0`` in its WHERE condition.
+3. The query users are being surprised by p90, p95, and p99 values that are taking into
+   account zero-duration click events which don't make sense for the use case.
+
+To resolve this confusion, we don't want to set quantilesState for ``click_duration`` if
+the incoming ``click_duration`` is 0.
+
+Steps to resolve the issue:
+
+1. Create a new materialized view ``click_events_aggregation_mv_v1`` via the migration system. This new materialized
+   view will use the WHERE clause or some kind of filtering to avoid setting quantilesState(0)
+   in the write for the ``click_duration`` column. This new materialized will only operate on
+   inputs in ``click_events_raw_local`` where ``materialization_version = 1``
+2. Test that this fixes the issue in your local environment by changing your consumer to use
+   ``materialization_version = 1``. It can make sense to control this via the settings file in
+   (in ``snuba/settings/__init.py__``)
+3. Run the migration in all relevant environments.
+4. Change the main consumer code, or the setting mentioned above in a specific environment, to
```

Comment:
Not a blocker for this PR, but something to consider. Would it make sense to control this via settings only going forward? Code changes and migrations would not always sync across different environments (SaaS, ST, Self-hosted). To solve this problem, a consumer needs to support all previous versions of writing to the raw table using the appropriate `materialization_version`. If this makes sense then we should recommend that you should make the change using settings.

Reply:
I think deprecation is complicated, but we actually do want to completely deprecate the old paths in the case of `metrics` and `generic_metrics` for example. I think it will be to our benefit if we can ensure that this happens in all environments where maintainers are keeping them up to date.

That said, I'm happy to change the recommendation here to be clearer and make it always settings-based because it's easy enough to switch code paths to something hard-coded